cli/azd/pkg/auth/cache.go

```diff
@@ -0,0 +1,31 @@
+package auth
+
+import (
+	"log"
+
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/cache"
+)
+
+// exportReplaceWithErrors is like cache.ExportReplace except each method is allowed to return an error.
+type exportReplaceWithErrors interface {
+	Replace(cache cache.Unmarshaler, key string) error
+	Export(cache cache.Marshaler, key string) error
+}
```

Comment:
This `Export`, `Replace` interface to replace the concept of `Write` and `Read` is definitely quite interesting.. It took a few tries to understand how `Replace` returns the result (which currently uses `cache.Unmarshaler` as a mechanism.

Is this a necessity that is imposed by the MSAL library? I didn't see explicit MSAL cache involved in the cache construction, so perhaps this is to set ourselves up to when that happens?

Personally, it seemed like having `Write(key string, data []byte) error` and `Read(key string) ([]byte, error)` would have made for an easier design.. With encryption simply transforming unencrypted to encrypted bytes before chaining the next `Write`. In this paradigm, I would also avoid calling the `Read` + `Write` interface a `cache` -- as the encryption passthrough strictly does not do caching, perhaps general `ReaderWriter`.

Or the other alternative is to avoid putting encryption as a middleware, and instead make it an option on `fileCache` to enable encryption. I personally like this the most.

Reply:
> Is this a necessity that is imposed by the MSAL library? I didn't see explicit MSAL cache involved in the cache construction, so perhaps this is to set ourselves up to when that happens?

Yes - a version of this interface (without the error returns) called `cache.ExportReplace` and msal libraries take this interface as the cache type when using the `public.WithCache` option.  In an earlier version of this PR we did not have this interface, we just implemented the version that didn't return errors and logged everything.

Since we needed to also save information for confidential clients (since MSAL does not provide a built-in way to do this) I figured we could re-use the design of the interface (but add an error return) and build an adapter to drop the errors and just log them for interfacing with MSAL.

But your right that this interface, as designed, is a bit of a mind bender.  I'll move all the logic into an interface like you describe and then build the MSAL adapter around that.  I think it will make the logic a little cleaner.

Reply:
This is all cleaned up now by adopting the interface you proposed.  I kept the middleware design for now instead of pushing the encryption stuff in to the file cache object.  I think this is going to set us up in the right direction.  Thanks for pushing on this, I'm much happier with the resulting code now.